### DISCUSSION TITLE
Protobuf Cue docs example and package naming confusion

### DISCUSSION THREAD

**Discussion by dylankb** [cue-lang/cue#1362] [General]
If you copy + paste the example for [extract a single .proto file](https://cuelang.org/docs/integrations/protobuf/#extract-cue-from-a-standalone-proto-file) into a VSCode IDE with `gopls` enabled, it will probably import the package `go/format` because it contains a [Node](https://pkg.go.dev/go/format#Node) function. However, what you really want is this package's [Node](https://pkg.go.dev/cuelang.org/go/cue/format#Node) function `cuelang.org/go/cue/format`.

1) Wondering if people have thought's on what might be an agreeable change to help next Cue newb:
a) Have the documentation examples include import statements
b) We could add a Go playground link, but obviously it won't run since there's no file system access (and it probably will time out downloading Cue related modules). 
c) Add a code comment. Less satisfying but very easy. 
d) Something else?

2) Kind of a random question aside, but any idea why Cue's package and method names match the Go standard library so closely in this case? Effective Go says we [shouldn't worry about package naming conflicts](https://golang.org/doc/effective_go#package-names), but in reality they can be a bit annoying to deal with.

**Comment by verdverm** [comment]
1. In the spirit of [reproducers](https://github.com/cue-lang/cue/wiki/Creating-test-or-performance-reproducers), I would vote to have examples which run without needing changes or additions.
2. CUE started as a fork of the Go code base (circa 2018 iirc) and Marcel was on the Go team until earlier this year. Go has a large influence on CUE implementation.

**Comment by myitcv** [comment]
@dylankb welcome to the CUE project!

> VSCode IDE with `gopls` enabled, it will probably import the package `go/format` because it contains a [Node](https://pkg.go.dev/go/format#Node) function

This is an unfortunate side effect of the behaviour of `go/import`, a library which is used as part of `gopls` but also forms the basis of a tool, `goimports`, originally used by editors as a `gofmt` equivalent but with [import-resolution support](https://pkg.go.dev/golang.org/x/tools/cmd/goimports). 

> a) Have the documentation examples include import statements

I think this is the best short-term solution. Please feel free to [submit a PR/CL](https://github.com/cue-lang/cue/blob/master/doc/contribute.md) to fix this. (I will note the contributing guide is going to get a revamp in the next couple of days to make it clearer, and it's also something we will be [live streaming about next week](https://twitter.com/cue_lang/status/1456623426496573440)). 

In the medium-long term:

* All the core documentation should be automatically tested as part of CI so that it doesn't rot. 
* Examples should, where possible, be interactive to give people a sense of how things work in practice. The playground is good for language-only examples that use the more primitive commands of `cmd/cue`, but we need a richer environment for any other tool examples (something like https://play-with-go.dev/)

As we discussed on [a recent community town hall](https://www.youtube.com/watch?v=GMTQcm3AwOU), all of these points roll up underneath a content strategy for CUE, which is one of the high priority items on our list.

**Comment by dylankb** [comment]
I agree with you both that having all documentation be testable and easily reproducible is the ideal state. 

Based on @myitcv's suggestion on how to move forward with a practical near term fix, though, I made tiny PR to address the specific issue I ran into [here](https://github.com/cue-lang/cuelang.org/pull/203).
